src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs

```diff
@@ -333,33 +338,38 @@ private IList<string> GetConfigFilePaths(ISettings settings)
         {
             var frameworkSorter = new NuGetFrameworkSorter();
 
+            var assetsFilePath = await GetAssetsFilePathAsync();
+            var fileInfo = new FileInfo(assetsFilePath);
+            PackageSpec assetsPackageSpec = default;
+            IList<LockFileTarget> targets = default;
+
+            if (fileInfo.Exists && fileInfo.LastWriteTimeUtc > _lastTimeAssetsModified)
+            {
+                await TaskScheduler.Default;
+                var lockFile = new LockFileFormat().Read(assetsFilePath);
+                assetsPackageSpec = lockFile.PackageSpec;
+                targets = lockFile.Targets;
+
+                _lastTimeAssetsModified = fileInfo.LastWriteTimeUtc;
+            }
+
             return packageSpec
-                .TargetFrameworks
-                .SelectMany(f => GetPackageReferences(f.Dependencies, f.FrameworkName))
-                .GroupBy(p => p.PackageIdentity)
-                .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First())
-                .ToArray();
+               .TargetFrameworks
+               .SelectMany(f => GetPackageReferences(f.Dependencies, f.FrameworkName, _installedPackages, assetsPackageSpec, targets))
+               .GroupBy(p => p.PackageIdentity)
+               .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First())
+               .ToArray();
         }
 
-        private static IEnumerable<PackageReference> GetPackageReferences(IEnumerable<LibraryDependency> libraries, NuGetFramework targetFramework)
+        private IEnumerable<PackageReference> GetPackageReferences(IEnumerable<LibraryDependency> libraries, NuGetFramework targetFramework, Dictionary<string, ProjectInstalledPackage> installedPackages, PackageSpec assetsPackageSpec, IList<LockFileTarget> targets)
         {
             return libraries
-                .Where(l => l.LibraryRange.TypeConstraint == LibraryDependencyTarget.Package)
-                .Select(l => ToPackageReference(l, targetFramework));
-        }
-
-        private static PackageReference ToPackageReference(LibraryDependency library, NuGetFramework targetFramework)
-        {
-            // The VersionRange can be null when the PackageReference items are for a project opted in the central package version management.
-            var identity = new PackageIdentity(
-                library.LibraryRange.Name,
-                library.LibraryRange.VersionRange?.MinVersion);
-
-            return new PackageReference(identity, targetFramework);
+                .Where(library => library.LibraryRange.TypeConstraint == LibraryDependencyTarget.Package)
+                .Select(library => new BuildIntegratedPackageReference(library, targetFramework, GetPackageReferenceUtility.UpdateResolvedVersion(library, targetFramework, assetsPackageSpec?.TargetFrameworks[0], targets, installedPackages)));
```

Comment:
Looks like we're guaranteed to have a TargetFrameworks[0] of at least Unsupported. Is there any concern with this breaking logic? Should we error-out here?

Reply:
why is this a breaking logic?

Reply:
@nkolev92 I asked if it could. There are a lot of utilities/etc that are called from this point. Do those work nicely with Unsupported? If not, there should be a check here.

I'm not sure that FirstOrDefault is a good change. Probably default(TargetFramework) is null, rather than Unsupported. And I'm still unclear how Unsupported might work from here on out.

Reply:
Ok this Equals will catch the null and return a null InstalledVersion. Seems we don't use this anywhere else, so maybe not an issue.
https://github.com/NuGet/NuGet.Client/pull/3539/files#diff-abdf4ad4e58bb17686322eb3506aa449R71